Contributing guide: add Quickstart and test details (#88)
cyberark/ansible-conjur-collection: CONTRIBUTING.md

    

 ```sh-session
 $ git clone https://github.com/cyberark/ansible-conjur-collection.git
 $ cd ansible-conjur-collection/dev
 $ ./start.sh

 ```
### Verification

  When start.sh script successfully setup Conjur environment along with inventory machines , the terminal returns the following:
        
   ```sh-session
   ...
   PLAY RECAP *********************************************************************
   ansibleplugingtestingconjurhostidentity-test_app_centos-1 : ok=17 ...
   ansibleplugingtestingconjurhostidentity-test_app_centos-2 : ok=17 ...
   ansibleplugingtestingconjurhostidentity-test_app_ubuntu-1 : ok=16 ...
   ansibleplugingtestingconjurhostidentity-test_app_ubuntu-2 : ok=16 ...
   
   ```

   After starting Conjur, your instance will be configured with the following:
   * Account: `cucumber`
   * User: `admin`
   * Password: Run `conjurctl role retrieve-key cucumber:user:admin` inside the Conjur container shell to retrieve the admin user API key (which is also the  password)

> ## Setup a Conjur OSS Environment

Author: Please share your suggestions on shared MS office doc , accordingly I will update here .

# Contributing to the Ansible Conjur Collection

> Thanks for your interest in Conjur Collection . Before contributing, please take a moment to

Author: Done

its license terms. Please email a signed copy to <a href="oss@cyberark.com">oss@cyberark.com</a>.
For general contribution and community guidelines, please see the [community repo](https://github.com/cyberark/community).

- [Contributing to the Ansible Conjur Collection](#contributing-to-the-ansible-conjur-collection)
  - [Prerequisites](#prerequisites)
  - [Set up a development environment](#set-up-a-development-environment)
      + [Setup a Conjur OSS Environment](#setup-a-conjur-oss-environment)
      + [Setup Conjur identity on managed host](#setup-conjur-identity-on-managed-host)
          - [Check Conjur identity](#check-conjur-identity)
          - [Set up Conjur identity](#set-up-conjur-identity)
          - [Set up Summon-Conjur](#set-up-summon-conjur)
   - [Testing](#testing)
   - [Releasing](#releasing)

 ## Prerequisites

To start developing and testing using our development scripts ,
the following tools need to be installed:

1. [Git][get-git] to manage source code
2. [Docker][get-docker] to manage dependencies and runtime environments
3. [Docker Compose][get-docker-compose] to orchestrate Docker environments
4. [Ruby][install-ruby] - native installation 

> 5. [Python][Python] - for lookup plug-in

Author: Done

[get-docker]: https://docs.docker.com/engine/installation
[get-docker-compose]: https://docs.docker.com/compose/install
[get-git]: https://git-scm.com/downloads
[install-ruby]: https://www.ruby-lang.org/en/documentation/installation/
[Python]: https://packaging.python.org/en/latest/tutorials/installing-packages/

## Set up a development environment

> For developers, it is easy-to-use development environment, so that they can work on the collection, roles, plugins, and playbooks without needing to run the test suite. The repo scripts is in dev/ to setup and teardown this environment.
> This docker-compose dev environment is really useful, including a few different services:
> -  A Conjur Open Source instance
> -  An Ansible control node
> -  Managed nodes to push tasks to
> 
> The `dev` directory contains a `docker-compose` file which creates a development
> environment with a database container (`pg`, short for *postgres*), and a
> `conjur` server container with source code mounted into the directory
> `/cyberark/dev/`.

Reviewer: I think we can consolidate this a bit:

```suggestion
The `dev` directory contains a `docker-compose` file which creates a development
environment with:
- a Conjur Open Source instance
- an Ansible control node
- a number of managed nodes
```

To use it:

1. Install dependencies (as above)

1. Start the container (and optional extensions):

   ```sh-session
   $ cd dev
   $ ./start.sh
   ...
   root@f75015718049:/cyberark/dev/#
   ```

   Once the `start` script finishes, you're in a Bash shell inside the Conjur
   server container.  To

   After starting Conjur, your instance will be configured with the following:
   * Account: `cucumber`
   * User: `admin`

>    * Password: Run `cat conjur.identity` inside the container shell to display the current logged-in identity (which is also the password)

Reviewer: Some of this section isn't accurate:

```md
1. Start the environment:

   ```sh-session
   $ cd dev
   $ ./start.sh
   ...
   PLAY RECAP *********************************************************************
   ansibleplugingtestingconjurhostidentity-test_app_centos-1 : ok=17 ...
   ansibleplugingtestingconjurhostidentity-test_app_centos-2 : ok=17 ...
   ansibleplugingtestingconjurhostidentity-test_app_ubuntu-1 : ok=16 ...
   ansibleplugingtestingconjurhostidentity-test_app_ubuntu-2 : ok=16 ...
   $
   \```

   Once the `start` script finishes, the environment's managed nodes have each been
   configured with a Conjur identity.

   After starting Conjur, your instance will be configured with the following:
   * Account: `cucumber`
   * User: `admin`
   * Password: Run `conjurctl role retrieve-key cucumber:user:admin` inside the Conjur container shell to retrieve the admin user API key (which is also the password)
```

1. Debug the server

   ```sh-session
   root@f39015718062:docker-compose exec <container-service-name> bash
   <various startup messages, then finally:>
   Use exit to stop

>    ```

Reviewer: This section can be removed - debugging Conjur isn't a focus here.

- copy files from the Ansible to the hosts  into /etc/Conjur.conf
- Request identity from Conjur
- Place identity file /etc/Conjur.identity when not Conjurized .

## Set up Summon-Conjur

- Download and unpack Summon
- Create folder for Summon-Conjur to be installed into
- Download and unpack Summon-Conjur

## Conjur Policy example

```sh
- !policy
  id: ansible
  annotations:
    description: Policy for Ansible master and remote hosts
  body:

  - !host
    id: ansible-master
    annotations:
      description: Host for running Ansible on remote targets

  - !layer &remote_hosts_layer
    id: remote_hosts
    annotations:
      description: Layer for Ansible remote hosts

  - !host-factory
    id: ansible-factory
    annotations:
      description: Factory to create new hosts for ansible
    layer: [ *remote_hosts_layer ]

  - !variable
    id: target-password
    annotations:
      description: Password needed by the Ansible remote machine

  - !permit
    role: *remote_hosts_layer
    privileges: [ execute ]
    resources: [ !variable target-password ]
```

> ## Useful links

Author: Done

| Source  | URLs |
| ------ | ------ |
| CyberArk Conjur |https://docs.conjur.org/Latest/en/Content/Integrations/ansible.html|
| GitHub | https://github.com/cyberark/ansible-conjur-collection|
| Ansible Galaxy | https://galaxy.ansible.com/cyberark/conjur_collection|

> | Ansible Doc| https://docs.ansible.com/ansible/latest/collections/cyberark/conjur/index.html|

Author: Done

# Contributing to the Ansible Conjur Collection

Thanks for your interest in Conjur. Before contributing, please take a moment to
read and sign our <a href="https://github.com/cyberark/community/blob/master/documents/CyberArk_Open_Source_Contributor_Agreement.pdf" download="conjur_contributor_agreement">Contributor Agreement</a>.
This provides patent protection for all Conjur users and allows CyberArk to enforce
its license terms. Please email a signed copy to <a href="oss@cyberark.com">oss@cyberark.com</a>.
For general contribution and community guidelines, please see the [community repo](https://github.com/cyberark/community).

> - Contributing to the Ansible Conjur Collection

Reviewer: ```suggestion
- [Contributing to the Ansible Conjur Collection](#contributing-to-the-ansible-conjur-collection)
```

  - [Prerequisites](#prerequisites)
  - [Set up a development environment](#set-up-a-development-environment)
  - [Testing Conjur OSS](#testing-conjur-oss)
  - [Testing Conjur Enterprise](#testing-conjur-enterprise)
  - [Generate conjur variable unit tests report](#generate-conjur-variable-unit-tests-report)
  - [Releasing](#releasing)
- [Ansible Conjur Collection Quick Start](#ansible-conjur-collection-quick-start)
    1. [Setup a conjur OSS Environment](#setup-a-conjur-oss-environment)
    2. [Load policy to set up Conjur Ansible integration](#load-policy-to-set-up-conjur-ansible-integration)
    3. [Create Ansible managed nodes](#create-ansible-managed-nodes)
    4. [Use `conjur_host_identity` to set up Conjur identity on managed nodes](#use-conjur-host-identity-to-set-up-conjur-identity-on-managed-nodes)
    5. [Use `conjur_variable` lookup plugin to provide secrets to Ansible Playbooks](#use-conjur-variable-lookup-plugin-to-provide-secrets-to-ansible-playbooks)

>  ## Prerequisites
> 
> To start developing and testing using our development scripts ,
> the following tools need to be installed:

Reviewer: ```suggestion
## Prerequisites

Before getting started, the following tools need to be installed:
```

1. [Git][get-git] to manage source code
2. [Docker][get-docker] to manage dependencies and runtime environments
3. [Docker Compose][get-docker-compose] to orchestrate Docker environments

[get-docker]: https://docs.docker.com/engine/installation
[get-docker-compose]: https://docs.docker.com/compose/install
[get-git]: https://git-scm.com/downloads

## Set up a development environment

The `dev` directory contains a `docker-compose` file which creates a development
environment :
-  A Conjur Open Source instance
-  An Ansible control node
-  Managed nodes to push tasks to

To use it:

1. Install dependencies (as above)

> 1. To setup the dev environment ,first need to Clone GitHub [conjur-collection](https://github.com/cyberark/ansible-conjur-collection) repository in your directory and then run start_oss.sh script

Reviewer: ```suggestion
1. To use the dev environment, clone the [Collection repository](https://github.com/cyberark/ansible-conjur-collection) and run the setup script:
```

 ```sh-session
 $ git clone https://github.com/cyberark/ansible-conjur-collection.git
 $ cd ansible-conjur-collection/dev
 $ ./start.sh

 ```
### Verification

>   When start_oss.sh script successfully setup Conjur environment along with inventory machines , the terminal returns the following:

Reviewer: ```suggestion
When the Conjur and Ansible containers have been successfully setup, the terminal prints the following:
```

   ```sh-session
   ...
   PLAY RECAP *********************************************************************
   ansibleplugingtestingconjurhostidentity-test_app_centos-1 : ok=17 ...
   ansibleplugingtestingconjurhostidentity-test_app_centos-2 : ok=17 ...
   ansibleplugingtestingconjurhostidentity-test_app_ubuntu-1 : ok=16 ...
   ansibleplugingtestingconjurhostidentity-test_app_ubuntu-2 : ok=16 ...

   ```

>    After starting Conjur, your instance will be configured with the following:

Reviewer: ```suggestion
Your Conjur instance will be configured with the following:
```

   * Account: `cucumber`
   * User: `admin`

>    * Password: Run `conjurctl role retrieve-key cucumber:user:admin` inside the Conjur container shell to retrieve the admin user API key (which is also the  password)

Reviewer: ```suggestion
   * Password: Run `conjurctl role retrieve-key cucumber:user:admin` inside the Conjur container shell to retrieve the admin user's API key
```

### Useful links

- [Official documentation for Conjur's Ansible integration](https://docs.conjur.org/Latest/en/Content/Integrations/ansible.html)
- [Conjur Collection on Ansible Galaxy](https://galaxy.ansible.com/cyberark/conjur)
- [Ansible documentation for the Conjur collection](https://docs.ansible.com/ansible/latest/collections/cyberark/conjur/index.html)

> ## Testing Conjur OSS

Reviewer: We should change the `Testing Conjur OSS`, `Testing Conjur Enterprise`, and `Generate conjur variable unit tests report` sections into these:

```
## Testing

### Unit tests

Unit tests are only available for the Conjur Variable Lookup plugin. To run these tests:

\```
./dev/test_unit.sh
\```

### Integration tests

The collection has integration tests for both the Variable Lookup plugin and the Host Identity role that will validate each against live Conjur and Ansible containers.

To run all tests:
\```
./ci/test.sh -a
\```

To run the tests for a particular module:
\```
./ci/test.sh -d <role or plugin name>
\```

Integration tests can be run against Conjur Enterprise by adding the `-e` flag:
\```
./ci/test/sh -e -a
\```
```

$ cd tests
$ ./test.sh -a
```

## Releasing

From a clean instance of main, perform the following actions to release a new version
of this plugin:

- Update the version number in [`galaxy.yml`](galaxy.yml) and [`CHANGELOG.md`](CHANGELOG.md)
    - Verify that all changes for this version in `CHANGELOG.md` are clear and accurate,
      and are followed by a link to their respective issue
    - Create a PR with these changes

- Create an annotated tag with the new version, formatted as `v##.##.##`
    - This will kick off an automated script which publish the release to
      [Ansible Galaxy](https://galaxy.ansible.com/cyberark/conjur)

- Create the release on GitHub for that tag
    - Build the release package with `./ci/build_release`
    - Attach package to Github Release

> 

Author: Done

## Set up Conjur Open Source and Ansible control node
  - ### Setup a conjur OSS Environment

    -   Pull the Docker image
    -   Generate the master key
    -   Load master key as an environment variable
    -   Start the Conjur OSS environment
    -   Create admin account

>     -   Connect the Conjur client to the Conjur server

Author: Done

          directly
      ```sh-session
      ansible-galaxy install cyberark.conjur-host-identity
      ```
      Once you've done this, you can configure each Ansible node with a Conjur identity by including a section like the example below in your Ansible playbook:

      ```sh-session
      - hosts: servers
        roles:
          - role: cyberark.conjur-host-identity
            conjur_appliance_url: 'https://conjur.myorg.com',
            conjur_account: 'myorg',
            conjur_host_factory_token: "{{lookup('env', 'HFTOKEN')}}",
            conjur_host_name: "{{inventory_hostname}}"
      ```
      First we register the host with Conjur, adding it into the layer specific to the provided host factory token, and then installs Summon with the Summon Conjur provider for secret retrieval from Conjur.

## Load policy to set up Conjur Ansible integration

  Policy defines Conjur entities and the relationships between them.  An entity can be a policy, a host, a user, a layer, a group, or a variable.

  ```sh
    docker exec conjur_client conjur policy load root /policy/root.yml
  ```
## Create Ansible managed nodes

>   Set all required variables, ensure `Conjur_host_factory_token` is set . Install `ca-certificates` and place Conjur public SSL certificate. Symlink Conjur public SSL certificate into /etc/ssl/certs .

Author: Done

> ## Use conjur host identity to set up Conjur identity on managed nodes

Reviewer: The content of this section should be replaced with the content of the section `### Granting a Conjur identity to Ansible hosts`. This section currently tells the user _how_ the role will configure their managed nodes - we should focus on usage instructions here.

Author: Done

  - Set all required variables, ensure `Conjur_host_factory_token` is set . Install `ca-certificates` , place Conjur  public SSL certificate and Symlink Conjur public SSL certificate into /etc/ssl/certs
  - Install openssl-perl Package when ansible_os_family is `RedHat`
  - Copy files from the Ansible to the hosts  into /etc/Conjur.conf and request identity from Conjur and place Conjur identity  file into /etc/

> ## Use conjur variable lookup plugin to provide secrets to Ansible Playbooks

Reviewer: This section should include a sample playbook that uses the lookup plugin. Something basic like this should do:
```yaml
---
- hosts: testapp
  tasks:
  - name: Provide secret with Lookup plugin
    debug:
      msg: "{{ lookup('cyberark.conjur.conjur_variable', '/ansible/target-password') }}"
```

Author: Done

    - Build the release package with `./ci/build_release`
    - Attach package to Github Release

# Ansible Conjur Collection Quick Start

## Setup a conjur OSS Environment

- Generate the master key, which will be used to encrypt Conjur's database. Store this value as an environment variable.

    ```sh-session
    docker-compose run --no-deps --rm conjur data-key generate > data_key
    export CONJUR_DATA_KEY="$(< data_key)"
    ```

- Start the Conjur OSS environment. An account, named `cucumber`, will be automatically created.

    ```sh-session
    docker-compose up -d conjur
    ```

- Retrieve the admin user's API key, and store the value in an environment variable.

    ```sh-session
    export CLI_CONJUR_AUTHN_API_KEY="$(docker-compose exec conjur conjurctl role retrieve-key cucumber:user:admin)"
    ```

- Start the Conjur CLI container. The CLI will be automatically authenticated as the user `cucumber:user:admin`.

    ```sh-session
    docker-compose up -d conjur-cli

>     ```

Author: Done

## Load policy to set up Conjur Ansible integration

  Policy defines Conjur entities and the relationships between them.  An entity can be a policy, a host, a user, a layer, a group, or a variable.

  Check out the policy file, and load it into Conjur:

  ```sh-session
  docker-compose exec conjur_cli cat /policy/root.yml
  docker-compose exec conjur_cli conjur policy load root /policy/root.yml
  ```

>   Also, load a dummy secret value into the 'ansible/target-password' variable. This is a variable required by remote nodes in order to complete their workloads.

Author: Done

  ```sh-session
  docker-compose exec conjur_cli conjur variable values add ansible/target-password S3cretV@lue
  ```
## Create Ansible managed nodes

  The Ansible environment will include a control node and a number of managed nodes. First, retrieve the API key for the Conjur host representing the control node, then create it:

  ```sh-session
  export ANSIBLE_CONJUR_AUTHN_API_KEY="$(docker-compose exec conjur conjurctl role retrieve-key cucumber:host:ansible/ansible-master)"
  docker-compose up -d ansible
  ```

  Next, create two instances of each managed node:

  ```sh-session
  docker-compose up -d --scale test_app_ubuntu=2 test_app_ubuntu
  docker-compose up -d --scale test_app_centos=2 test_app_centos
  ```

> ## Use conjur host identity to set up Conjur identity on managed nodes

Reviewer: ```suggestion
## Use Conjur Ansible Role to set up identity on managed nodes
```

Author: Done

>   To grant your Ansible host a Conjur identity, you first must install the Conjur Ansible Role in your playbook directly
> 
>   ```sh-session
>   ansible-galaxy install cyberark.conjur-host-identity
>   ```

Reviewer: ```suggestion
  To grant your Ansible host a Conjur identity, first install the Conjur Collection on your Ansible control node:

  \```sh-session
  ansible-galaxy collection install cyberark.conjur
  \```
```

Author: Done

  Once you've done this, you can configure each Ansible node with a Conjur identity by including a section like the example below in your Ansible playbook:

  ```sh-session
  - hosts: servers
    roles:

>       - role: cyberark.conjur-host-identity

Reviewer: ```suggestion
      - role: cyberark.conjur.conjur-host-identity
```

Author: Done

        conjur_appliance_url: 'https://conjur.myorg.com',
        conjur_account: 'myorg',

>         conjur_host_factory_token: "{{lookup('env', 'HFTOKEN')}}",

Reviewer: We need an additional step before this one setting up the host factory token in the `HFTOKEN` env var

Author: Done

        conjur_host_name: "{{inventory_hostname}}"
  ```
  First we register the host with Conjur, adding it into the layer specific to the provided host factory token, and then installs Summon with the Summon Conjur provider for secret retrieval from Conjur.

> ## Use conjur variable lookup plugin to provide secrets to Ansible Playbooks

Reviewer: ```suggestion
## Use Conjur Ansible Lookup plugin to provide secrets to Ansible Playbooks
```

Author: Done
